src/api/Api.js

```diff
@@ -95,65 +157,142 @@ export class Api {
+  setBaseUrl (baseUrl) {
+    this.baseUrl = baseUrl
+  }
+
+  /* #region request methods */
+
+  _compileUrl (url, params) {
+    let urlObj = null
+    const baseUrlObj = new URL(this.baseUrl)
+
+    if (url.indexOf('//') === 0) {
+      url = baseUrlObj.protocol + url
+    }
 
     if (!isAbsoluteUri(url)) {
-      url = this.baseUrl + url
+      // various path adjustments
+      url = url.replace(/^(\/)(.+)/, '$2')
+
+      if (baseUrlObj.pathname !== '' && baseUrlObj.pathname !== '/') {
+        const deSlashedBaseUrlPath = baseUrlObj.pathname.replace(
+          /(\/)(.+)(\/)$/,
+          '$2'
+        )
+
+        url = deSlashedBaseUrlPath + '/' + url
+
+        baseUrlObj.pathname = ''
+      }
+
+      urlObj = new URL(baseUrlObj.href)
+      urlObj.pathname = url
+    } else {
+      try {
+        urlObj = new URL(url)
+      } catch (e) {
+        throw new ApiError('Invalid url: ' + url)
+      }
+    }
+
+    urlObj.search = stringify(params, { encodeValuesOnly: true, arrayFormat: 'brackets' })
+
+    return urlObj.href
+  }
+
+  isUrlCrossDomain (url) {
+    if (typeof url !== 'string') {
+      return false
+    }
+
+    if (url.length === 0) {
+      return false
     }
 
-    // make cross domain requests if necessary
-    let crossDomain = false
-    if (url.length > 0 && isAbsoluteUri(url)) {
-      crossDomain = true
+    if (url.indexOf('//') === 0) {
+      // this is only valid in here as the url is not actually used this way!
+      // please check _compileUrl for the correct way to do this
+      url = 'http:' + url
     }
 
-    return axios.create(config)
-      .request({ method, url, params, data, crossDomain })
-      .then(
-        async response => {
-          for (let i = 0; i < this.preprocessingCallbacks.length; i++) {
-            await this.preprocessingCallbacks[i](response)
-          }
+    try {
+      const urlObj = new URL(url)
+      const baseUrlObj = new URL(this.baseUrl)
 
-          return response
-        }
-      ).catch(
-        async errorResponse => {
-          for (let i = 0; i < this.errorCallbacks.length; i++) {
-            await this.errorCallbacks[i](errorResponse)
-          }
+      return urlObj.host !== baseUrlObj.host
+    } catch (e) {}
 
-          return errorResponse
-        }
-      )
+    return url.length > 0 && isAbsoluteUri(url)
   }
 
-  get (url, params = null, options = {}) {
-    return this.doRequest('get', url, params, null, options)
+  /**
+   *
+   * @param {string} method
+   * @param {string} url
+   * @param {Object} params
+   * @param {Object} data
+   * @returns {Promise<Response>}
+   * @protected
+   */
+  async _doRequest (method, url, params, data) {
+    url = this._compileUrl(url, params)
+
+    const requestConfig = {
+      body: null,
+      headers: this.headers,
+      method: method.toUpperCase(),
+      mode: this.isUrlCrossDomain(url) ? 'cors' : 'same-origin',
+      referrerPolicy: 'strict-origin-when-cross-origin'
```

Comment:
to be debated, I went with the middle way between privacy aware but user friendly for now